src/Microsoft.AspNetCore.Server.Kestrel.Core/Adapter/Internal/AdaptedPipeline.cs

```diff
@@ -7,73 +7,131 @@
 using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http;
 using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure;
 using Microsoft.AspNetCore.Server.Kestrel.Internal.System.IO.Pipelines;
-using Microsoft.Extensions.Logging;
 
 namespace Microsoft.AspNetCore.Server.Kestrel.Core.Adapter.Internal
 {
     public class AdaptedPipeline
     {
         private const int MinAllocBufferSize = 2048;
 
-        private readonly Stream _filteredStream;
-        private readonly StreamSocketOutput _output;
+        private readonly Stream _stream;
         private readonly IKestrelTrace _trace;
+        private readonly IPipeWriter _transportOutputPipe;
+        private readonly IPipeReader _transportInputPipe;
 
         public AdaptedPipeline(
-            Stream filteredStream,
+            Stream stream,
+            IPipeReader transportInputPipe,
+            IPipeWriter transportOutputPipe,
             IPipe inputPipe,
             IPipe outputPipe,
             IKestrelTrace trace)
         {
+            _stream = stream;
+            _transportInputPipe = transportInputPipe;
+            _transportOutputPipe = transportOutputPipe;
             Input = inputPipe;
-            _output = new StreamSocketOutput(filteredStream, outputPipe);
-            _filteredStream = filteredStream;
+            Output = outputPipe;
             _trace = trace;
         }
 
         public IPipe Input { get; }
 
-        public ISocketOutput Output => _output;
+        public IPipe Output { get; }
 
         public async Task RunAsync()
         {
+            var inputTask = ReadInputAsync();
+            var outputTask = WriteOutputAsync();
+
+            await inputTask;
+            await outputTask;
+        }
+
+        private async Task WriteOutputAsync()
+        {
             try
             {
-                var inputTask = ReadInputAsync();
-                var outputTask = _output.WriteOutputAsync();
+                while (true)
+                {
+                    var readResult = await Output.Reader.ReadAsync();
+                    var buffer = readResult.Buffer;
 
-                await inputTask;
+                    try
+                    {
+                        if (buffer.IsEmpty && readResult.IsCompleted)
+                        {
+                            break;
+                        }
 
-                _output.Dispose();
+                        if (buffer.IsEmpty)
+                        {
+                            await _stream.FlushAsync();
+                        }
```

Comment:
Since this is a hot path, optimize the common double check?
```c#
if (buffer.IsEmpty)
{
    if (readResult.IsCompleted)
    {
        break;
    }
    await _stream.FlushAsync();
}
```